### PR TITLE
Iskeyword match

### PR DESCRIPTION
**Pull request by kawas44**
Hi,

As I was using `ddc-buffer` on programming source files, I found that the candidates were not respecting the programming language `iskeyword` option.

I would like to submit this PR to make use of `iskeyword`. Most of this PR is inspired by `Shougo/ddc-around`.

Don't hesitate to modify/discuss/merge at your will.

Cheers
Kawas

### CODE REVIEW

**Review by matsui54 — CHANGES_REQUESTED** [review]

File: denops/@ddc-sources/buffer.ts
Comment (kawas44) [reply denops/@ddc-sources/buffer.ts:66]: I am not sure that we should use `sourceOptions.keywordPattern`, which is the pattern for the current edited buffer only.

Imagine we are editing a Clojure file and we have in split windows a Java file or an SQL file or a Markdown file, each buffer has it's own iskeyword option that make sense in their file type. Do we want to collect words of the other buffers with the Clojure ([rather wide](https://github.com/clojure-vim/clojure.vim/blob/b54a4d6a01950398330090683d5b304cc8a5f71f/ftplugin/clojure.vim#L20)) iskeyword ?

Comment (matsui54) [reply denops/@ddc-sources/buffer.ts:66]: `iskeyword` from `op.iskeyword.getLocal()` is also local to the current edited buffer, so collecting words with inappropriate `iskeyword` can't be avoided.

I think this is not so big problem because ddc-buffer rarely gathers words from other buffers .
`makeCurrentBufCache()`, which gathers words from other buffers, is called when there is no cache for the buffers to gather words, and this happens only on initialization.

Comment (kawas44) [reply denops/@ddc-sources/buffer.ts:66]: Ok, I understand. Will update the PR accordingly.

Comment (kawas44) [reply denops/@ddc-sources/buffer.ts]: Hi Matsui54,

I just found out that we can get the `iskeyword` buffer local value using denops api `fn.getbufvar`.

I have tested it successfully, by opening a markdown and a typescript file in the same tab and then enter insert mode to trigger onInit and onEvent methods. Here are the logs, tracing events and `iskeyword` resolution (buffer 4 is the typescript file which has a longer pattern).

2022-01-19_ddc-buffer_test-iskeyword

Would that be a better implementation?

Comment (matsui54) [reply denops/@ddc-sources/buffer.ts]: Great! But at the same time, we should consider `sourceOptions.keywordPattern` as possible because users may want to override `iskeyword` by this ddc-option.
How about doing like this?
```typescript
    let pattern;
    if (bufnr == (await fn.bufnr(denops))) {
      // current buffer
      pattern = sourceOptions.keywordPattern;
    } else {
      // other buffer
      const iskeyword = await fn.getbufvar(denops, bufnr, "&iskeyword") as string
      pattern = "[" + vimoption2ts(iskeyword) + "]*";
    }
```

Comment (kawas44) [reply denops/@ddc-sources/buffer.ts]: Ok, I will try to do something in this way.

**Comment by kawas44** [comment]
Hi matsui54,

After playing with the code, analysing when the source is initialized and which events trigger the source (and initialization if not already done...), I have made some changes to use ddc `options.keywordPattern`.

Because of the way events trigger a *current buffer* indexation, in most case we only need to use ddc `options.keywordPattern` (as you told me at the very beginning of our conversation).

As always, waiting for your remarks and ready to squash stuff before a merge :)

**Comment by matsui54** [comment]
Tested the PR and it seems perfect! Thank you for your contribution!

**Pull request merged**
